File: src/topic03_perception/cpp/video_publisher.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <sensor_msgs/image_encodings.h>
#include <std_msgs/Header.h>
#include <stdio.h>

using namespace std;
using namespace cv;

int main (int argc, char** argv)
{
    ros::init(argc, argv, "video_publisher_node");
    ros::NodeHandle nh;
    image_transport::ImageTransport it(nh);
    image_transport::Publisher image_pub = it.advertise("/video", 1);

    // Import video
    VideoCapture video("/home/daeronhing/Desktop/ROS_OpenCV/catkin_ws/src/ros_essentials_cpp-ros-noetic/src/topic03_perception/video/tennis-ball-video.mp4");

    Mat frame;
    while(true) {
        video >> frame;

        // Publish the image
        sensor_msgs::ImagePtr imgmsg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();

        image_pub.publish(imgmsg);
        ros::spinOnce();

        if(waitKey(30) >= 0) break;
    }

    ROS_INFO("Finish");
    return 0;
}

File: src/topic03_perception/video_publisher.py
#!/usr/bin/env python3

import numpy as np
import rospy
import cv2
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

def import_video(src):
    video = cv2.VideoCapture(src)
    return video

def frame_rate(hz):
    return (int)(1000/hz)

if __name__ == "__main__":
    bridge = CvBridge()
    
    # Declare video path
    video_path = "/dev/video0"  # webcam
    # video_path = "/home/daeronhing/Desktop/ROS_OpenCV/catkin_ws/src/ros_essentials_cpp-ros-noetic/src/topic03_perception/video/tennis-ball-video.mp4"

    video = import_video(video_path)
    
    rospy.init_node("video_publisher")
    video_publisher = rospy.Publisher("/video", Image)
    
    rospy.loginfo("Video publisher node started")
    
    while(True):
        ret, frame = video.read()
        
        try:
            print(ret)
            cv2.imshow("Test Publisher", frame)
            video_publisher.publish(bridge.cv2_to_imgmsg(frame, "bgr8"))
            # bridge.cv2_to_imgmsg(frame, 'bgr8')
            
        except CvBridgeError as e:
            print(e)
        
        if cv2.waitKey(frame_rate(10)) & 0xFF==ord('q'): 
            break
        
    video.release()
    cv2.destroyAllWindows()